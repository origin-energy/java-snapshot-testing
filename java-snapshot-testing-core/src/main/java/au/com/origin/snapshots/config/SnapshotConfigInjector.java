package au.com.origin.snapshots.config;

public interface SnapshotConfigInjector {
  SnapshotConfig getSnapshotConfig();
}
