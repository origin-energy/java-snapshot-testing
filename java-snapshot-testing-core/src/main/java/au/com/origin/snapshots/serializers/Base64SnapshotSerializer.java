package au.com.origin.snapshots.serializers;

import au.com.origin.snapshots.Snapshot;
import au.com.origin.snapshots.SnapshotSerializerContext;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Base64;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * This Serializer converts a byte[] into a base64 encoded string.  If the input is not
 * a byte[] it will be converted using `.getBytes(StandardCharsets.UTF_8)` method
 */
public class Base64SnapshotSerializer implements SnapshotSerializer {
  private static final ToStringSnapshotSerializer toStringSnapshotSerializer =
      new ToStringSnapshotSerializer();

  @Override
  public Snapshot apply(Object[] objects, SnapshotSerializerContext gen) {
    List<?> encoded = Arrays.stream(objects)
        .filter(Objects::nonNull)
        .map(it -> {
          byte[] bytes = it instanceof byte[] ? (byte[]) it : it.toString().getBytes(StandardCharsets.UTF_8);
          return Base64.getEncoder().encodeToString(bytes);
        })
        .collect(Collectors.toList());
    return toStringSnapshotSerializer.apply(encoded.toArray(), gen);
  }

  @Override
  public String getOutputFormat() {
    return SerializerType.BASE64.name();
  }
}
