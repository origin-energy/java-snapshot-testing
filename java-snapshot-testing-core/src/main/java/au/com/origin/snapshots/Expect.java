package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RequiredArgsConstructor
public class Expect {
  private final SnapshotVerifier snapshotVerifier;
  private final Method testMethod;

  private SnapshotSerializer snapshotSerializer;
  private SnapshotComparator snapshotComparator;
  private List<SnapshotReporter> snapshotReporters;
  private String scenario;

  private final Map<String, String> headers = new HashMap<>();

  public static Expect of(SnapshotVerifier snapshotVerifier, Method method) {
    return new Expect(snapshotVerifier, method);
  }

  /**
   * Make an assertion on the given input parameters against what already exists
   *
   * @param firstObject first snapshot object
   * @param objects     other snapshot objects
   */
  public void toMatchSnapshot(Object firstObject, Object... objects) {
    SnapshotContext snapshotContext = snapshotVerifier.expectCondition(testMethod, firstObject, objects);
    if (snapshotSerializer != null) {
      snapshotContext.setSnapshotSerializer(snapshotSerializer);
    }
    if (snapshotComparator != null) {
      snapshotContext.setSnapshotComparator(snapshotComparator);
    }
    if (snapshotReporters != null) {
      snapshotContext.setSnapshotReporters(snapshotReporters);
    }
    if (scenario != null) {
      snapshotContext.setScenario(scenario);
    }
    snapshotContext.header.putAll(headers);

    snapshotContext.toMatchSnapshot();
  }


  /**
   * Normally a snapshot can be applied only once to a test method.
   * <p>
   * For Parameterized tests where the same method is executed multiple times you can supply
   * the scenario() to overcome this restriction.  Ensure each scenario is unique.
   *
   * @param scenario - unique scenario description
   * @return Snapshot
   */
  public Expect scenario(String scenario) {
    this.scenario = scenario;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param serializer your custom serializer
   * @return Snapshot
   */
  public Expect serializer(SnapshotSerializer serializer) {
    this.snapshotSerializer = serializer;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param name - the {name} attribute serializer.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect serializer(String name) {
    this.snapshotSerializer = SnapshotProperties.getInstance("serializer." + name);
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param comparator your custom comparator
   * @return Snapshot
   */
  public Expect comparator(SnapshotComparator comparator) {
    this.snapshotComparator = comparator;
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param name the {name} attribute comparator.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect comparator(String name) {
    this.snapshotComparator = SnapshotProperties.getInstance("comparator." + name);
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot
   * This will replace the default reporters defined in the config
   *
   * @param reporters your custom reporters
   * @return Snapshot
   */
  public Expect reporters(SnapshotReporter... reporters) {
    this.snapshotReporters = Arrays.asList(reporters);
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot
   * This will replace the default reporters defined in the config
   *
   * @param name the {name} attribute reporters.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect reporters(String name) {
    this.snapshotReporters = SnapshotProperties.getInstances("reporters." + name);
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot.
   *
   * @param serializer your custom serializer
   * @return this
   * @see au.com.origin.snapshots.serializers.SnapshotSerializer
   * <p>
   * Example implementations
   * @see au.com.origin.snapshots.serializers.ToStringSnapshotSerializer
   * @see au.com.origin.snapshots.serializers.Base64SnapshotSerializer
   */
  @SneakyThrows
  public Expect serializer(Class<? extends SnapshotSerializer> serializer) {
    this.snapshotSerializer = serializer.getConstructor().newInstance();
    return this;
  }

  /**
   * Add anything you like to the snapshot header.
   *
   * These custom headers can be used in serializers, comparators or reporters to change
   * how they behave.
   *
   * @param key key
   * @param value value
   * @return Expect
   */
  public Expect header(String key, String value) {
    headers.put(key, value);
    return this;
  }

}
