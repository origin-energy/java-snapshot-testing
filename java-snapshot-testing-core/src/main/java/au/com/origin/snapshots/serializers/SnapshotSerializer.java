package au.com.origin.snapshots.serializers;

import au.com.origin.snapshots.Snapshot;
import au.com.origin.snapshots.SnapshotSerializerContext;

import java.util.function.BiFunction;

public interface SnapshotSerializer extends BiFunction<Object[], SnapshotSerializerContext, Snapshot> {
  String getOutputFormat();
}
