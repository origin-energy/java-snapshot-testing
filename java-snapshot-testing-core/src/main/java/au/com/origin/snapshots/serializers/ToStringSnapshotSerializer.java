package au.com.origin.snapshots.serializers;

import au.com.origin.snapshots.Snapshot;
import au.com.origin.snapshots.SnapshotFile;
import au.com.origin.snapshots.SnapshotSerializerContext;
import lombok.extern.slf4j.Slf4j;

import java.util.Arrays;
import java.util.stream.Collectors;

/**
 * This Serializer does a snapshot of the {@link Object#toString()} method
 * <p>
 * Will render each toString() on a separate line
 */
@Slf4j
public class ToStringSnapshotSerializer implements SnapshotSerializer {

  @Override
  public Snapshot apply(Object[] objects, SnapshotSerializerContext gen) {
    String body = "[\n" + Arrays.stream(objects)
        .map(Object::toString)
        .map(it -> {
          if (it.contains(SnapshotFile.SPLIT_STRING)) {
            log.warn("Found 3 consecutive lines in your snapshot \\n\\n\\n. This sequence is reserved as the snapshot separator - replacing with \\n.\\n.\\n");
            return it.replaceAll(SnapshotFile.SPLIT_STRING, "\n.\n.\n");
          }
          return it;
        })
        .collect(Collectors.joining("\n")) +
        "\n]";
      return gen.toSnapshot(body);
  }

  @Override
  public String getOutputFormat() {
    return SerializerType.TEXT.name();
  }
}
