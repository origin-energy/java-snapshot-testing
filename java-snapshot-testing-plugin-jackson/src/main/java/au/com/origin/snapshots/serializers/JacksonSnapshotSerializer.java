package au.com.origin.snapshots.serializers;

import au.com.origin.snapshots.Snapshot;
import au.com.origin.snapshots.SnapshotSerializerContext;
import au.com.origin.snapshots.exceptions.SnapshotExtensionException;
import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.PrettyPrinter;
import com.fasterxml.jackson.core.util.DefaultIndenter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import com.fasterxml.jackson.core.util.Separators;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

public class JacksonSnapshotSerializer implements SnapshotSerializer {

  private final PrettyPrinter pp = new DefaultPrettyPrinter("") {
    {
      Indenter lfOnlyIndenter = new DefaultIndenter("  ", "\n");
      this.indentArraysWith(lfOnlyIndenter);
      this.indentObjectsWith(lfOnlyIndenter);
    }

    // It's a requirement
    // @see https://github.com/FasterXML/jackson-databind/issues/2203
    public DefaultPrettyPrinter createInstance() {
      return new DefaultPrettyPrinter(this);
    }

    @Override
    public DefaultPrettyPrinter withSeparators(Separators separators) {
      this._separators = separators;
      this._objectFieldValueSeparatorWithSpaces =
          separators.getObjectFieldValueSeparator() + " ";
      return this;
    }
  };
  private final ObjectMapper objectMapper = new ObjectMapper() {{
    this.enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS);
    this.enable(SerializationFeature.WRITE_DATES_WITH_ZONE_ID);
    this.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
    this.disable(SerializationFeature.FAIL_ON_EMPTY_BEANS);
    this.setSerializationInclusion(JsonInclude.Include.NON_NULL);

    this.registerModule(new JavaTimeModule());
    this.registerModule(new Jdk8Module());

    this.setVisibility(
        this
            .getSerializationConfig()
            .getDefaultVisibilityChecker()
            .withFieldVisibility(JsonAutoDetect.Visibility.ANY)
            .withGetterVisibility(JsonAutoDetect.Visibility.NONE)
            .withSetterVisibility(JsonAutoDetect.Visibility.NONE)
            .withCreatorVisibility(JsonAutoDetect.Visibility.NONE));
    JacksonSnapshotSerializer.this.configure(this);
  }};

  /**
   * Override to customize the Jackson objectMapper
   *
   * @param objectMapper existing ObjectMapper
   */
  public void configure(ObjectMapper objectMapper) {
  }

  @Override
  public Snapshot apply(Object[] objects, SnapshotSerializerContext gen) {
    try {
      String body = objectMapper.writer(pp).writeValueAsString(objects);
      return  gen.toSnapshot(body);
    } catch (Exception e) {
      throw new SnapshotExtensionException("Jackson Serialization failed", e);
    }
  }

  @Override
  public String getOutputFormat() {
    return SerializerType.JSON.name();
  }
}
