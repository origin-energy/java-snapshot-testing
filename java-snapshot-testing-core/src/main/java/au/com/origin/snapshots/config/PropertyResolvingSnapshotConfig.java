package au.com.origin.snapshots.config;

import au.com.origin.snapshots.SnapshotProperties;
import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;

import java.util.List;

public class PropertyResolvingSnapshotConfig implements SnapshotConfig {

  @Override
  public String getOutputDir() {
    return SnapshotProperties.getOrThrow("output-dir");
  }

  @Override
  public String getSnapshotDir() {
    return SnapshotProperties.getOrThrow("snapshot-dir");
  }

  @Override
  public SnapshotSerializer getSerializer() {
    return SnapshotProperties.getInstance("serializer");
  }

  @Override
  public SnapshotComparator getComparator() {
    return SnapshotProperties.getInstance("comparator");
  }

  @Override
  public List<SnapshotReporter> getReporters() {
    return SnapshotProperties.getInstances("reporters");
  }

  @Override
  public boolean isCI() {
    String envVariable = SnapshotProperties.getOrThrow("ci-env-var");
    return System.getenv(envVariable) != null;
  }
}
