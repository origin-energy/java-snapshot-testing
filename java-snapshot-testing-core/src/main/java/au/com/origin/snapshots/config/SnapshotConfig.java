package au.com.origin.snapshots.config;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;

import java.util.List;
import java.util.Optional;

/**
 * Snapshot Configuration
 * ----------------------
 * <p>
 * Implement this interface when integrating `java-snapshot-testing` with a custom testing library
 */
public interface SnapshotConfig {
  String JVM_UPDATE_SNAPSHOTS_PARAMETER = "updateSnapshot";

  /**
   * The base directory where files get written (excluding package directories)
   * default: "src/test/java"
   * <p>
   * You might want to override if you have tests under "src/test/integration" for example
   *
   * @return snapshot output folder
   */
  String getOutputDir();

  /**
   * Subdirectory to store snapshots in
   *
   * @return name of subdirectory
   */
  String getSnapshotDir();

  /**
   * Optional
   *
   * @return snapshots should be updated automatically without verification
   */
  default Optional<String> updateSnapshot() {
    return Optional.ofNullable(System.getProperty(JVM_UPDATE_SNAPSHOTS_PARAMETER));
  }

  /**
   * Optional
   * Override to supply your own custom serialization function
   *
   * @return custom serialization function
   */
  SnapshotSerializer getSerializer();

  /**
   * Optional
   * Allows you to perform any custom modifications to the file before it is saved
   *
   * @param testClass       target test class
   * @param snapshotContent snapshot file as a string
   * @return snapshot file contents to be persisted
   */
  default String onSaveSnapshotFile(Class<?> testClass, String snapshotContent) {
    return snapshotContent;
  }

  /**
   * Optional
   * Override to supply your own custom comparator function
   *
   * @return custom comparator function
   */
  SnapshotComparator getComparator();

  /**
   * Optional
   * Override to supply your own custom reporter functions
   * Reporters will run in the same sequence as provided.
   * Reporters should throw exceptions to indicate comparison failure.
   * Exceptions thrown from reporters are aggregated and reported together.
   * Reporters that wish to leverage IDE comparison tools can use standard
   * assertion libraries like assertj, junit jupiter assertions (or) opentest4j.
   *
   * @return custom reporter functions
   */
  List<SnapshotReporter> getReporters();

  /**
   * Optional
   * This method is meant to detect if we're running on a CI environment.
   * This is used to determine the action to be taken when a snapshot is not found.
   * <p>
   * If this method returns false, meaning we're NOT running on a CI environment (probably a dev machine),
   * a new snapshot is created when not found.
   * <p>
   * If this method returns true, meaning we're running on a CI environment, no new snapshots are created
   * and an error is thrown instead to prevent tests from silently passing when snapshots are not found.
   * <p>
   * Often to determine if running on a CI environment is to check for the presence of a 'CI' env variable
   *
   * @return boolean indicating if we're running on a CI environment or not
   */
  boolean isCI();

}
