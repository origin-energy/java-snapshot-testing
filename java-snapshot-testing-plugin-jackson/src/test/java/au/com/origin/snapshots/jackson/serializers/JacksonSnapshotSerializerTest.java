package au.com.origin.snapshots.jackson.serializers;

import au.com.origin.snapshots.*;
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig;
import au.com.origin.snapshots.config.SnapshotConfig;
import au.com.origin.snapshots.serializers.SerializerType;
import au.com.origin.snapshots.serializers.SnapshotSerializer;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.util.*;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;

public class JacksonSnapshotSerializerTest {

  private static final SnapshotConfig DEFAULT_CONFIG =
      new PropertyResolvingSnapshotConfig() {
        @Override
        public SnapshotSerializer getSerializer() {
          return new JacksonSnapshotSerializer();
        }
      };

  private SnapshotSerializerContext gen =
      new SnapshotSerializerContext(
          "test", null, new SnapshotHeader(), JacksonSnapshotSerializerTest.class, null);

  @Test
  public void shouldSerializeMap() {
    Map<String, Object> map = new HashMap<>();
    map.put("name", "John Doe");
    map.put("age", 40);

    SnapshotSerializer serializer = new JacksonSnapshotSerializer();
    Snapshot result = serializer.apply(map, gen);
    Assertions.assertThat(result.getBody())
        .isEqualTo(
            "[\n"
                + "  {\n"
                + "    \"age\": 40,\n"
                + "    \"name\": \"John Doe\"\n"
                + "  }\n"
                + "]");
  }

  @Test
  public void shouldSerializeDifferentTypes(TestInfo testInfo) {
    SnapshotVerifier snapshotVerifier =
        new SnapshotVerifier(DEFAULT_CONFIG, testInfo.getTestClass().get());
    Expect expect = Expect.of(snapshotVerifier, testInfo.getTestMethod().get());
    expect.toMatchSnapshot(new TypeDummy());
    snapshotVerifier.validateSnapshots();
  }

  @Test
  void shouldSupportJsonFormat() {
    Assertions.assertThat(new JacksonSnapshotSerializer().getOutputFormat())
        .isEqualTo(SerializerType.JSON.name());
  }

  private Map<String, Integer> deterministicMap(Map<String, Integer> target) {
    final List<String> items =
        new ArrayList<String>() {
          {
            add("f");
            add("a");
            add("d");
            add("e");
            add("g");
            add("b");
            add("c");
          }
        };
    items.forEach(it -> target.put(it, (int) it.charAt(0)));
    return target;
  }

  private Collection<String> deterministicCollection(Collection<String> target) {
    final List<String> items =
        new ArrayList<String>() {
          {
            add("f");
            add("a");
            add("d");
            add("e");
            add("g");
            add("b");
            add("c");
          }
        };
    target.addAll(items);
    return target;
  }

  private enum AnEnum {
    F,
    A,
    D,
    E,
    G,
    B,
    C;
  }

  private final class TypeDummy {
    private final Void aNull = null;
    private final Object anObject = new Object();
    private final byte aByte = "A".getBytes()[0];
    private final short aShort = 32767;
    private final int anInt = 2147483647;
    private final long aLong = 9223372036854775807L;
    /**
     * Note: Stripping trailing zeroes is required to force {@link BigDecimal#toString()} to use scientific notation.
     */
    private final BigDecimal aBigDecimal = BigDecimal.valueOf(100).stripTrailingZeros();
    private final float aFloat = 0.1234567F;
    private final double aDouble = 1.123456789123456D;
    private final boolean aBoolean = true;
    private final char aChar = 'A';
    private final String string = "Hello World";
    private final Date date = Date.from(Instant.parse("2020-10-19T22:21:07.103Z"));
    private final LocalDate localDate = LocalDate.parse("2020-10-19");
    private final LocalDateTime localDateTime = LocalDateTime.parse("2020-10-19T22:21:07.103");
    private final ZonedDateTime zonedDateTime =
        ZonedDateTime.parse("2020-04-19T22:21:07.103+10:00[Australia/Melbourne]");
    private final AnEnum anEnum = AnEnum.A;
    private final Optional<String> presentOptional = Optional.of("Hello World");
    private final Optional<String> emptyOptional = Optional.empty();
    private final String[] stringArray = {"f", "a", "d", "e", "g", "b", "c"};
    private final Object[] anEnumArray = Arrays.stream(AnEnum.values()).toArray();

    // Maps
    private final Map<String, Integer> hashMap = deterministicMap(new HashMap<>());
    private final Map<String, Integer> treeMap = deterministicMap(new TreeMap<>());
    private final Map<String, Integer> linkedHashMap = deterministicMap(new LinkedHashMap<>());

    // Sets
    private final Collection<String> linkedHashSet = deterministicCollection(new LinkedHashSet<>());
    private final Collection<String> hashSet = deterministicCollection(new HashSet<>());
    private final Collection<String> treeSet = deterministicCollection(new TreeSet<>());

    // Lists
    private final Collection<String> arrayList = deterministicCollection(new ArrayList<>());
    private final Collection<String> linkedList = deterministicCollection(new LinkedList<>());

    // Mixed Maps, Sets, Lists
    private final Collection<Object> listOfCollections =
        new ArrayList<Object>() {
          {
            add(deterministicMap(new LinkedHashMap<>()));
            add(deterministicCollection(new LinkedHashSet<>()));
            add(deterministicCollection(new LinkedList<>()));
          }
        };
  }
}
