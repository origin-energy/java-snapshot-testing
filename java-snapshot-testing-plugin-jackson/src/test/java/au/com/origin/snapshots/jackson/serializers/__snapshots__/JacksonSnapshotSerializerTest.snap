au.com.origin.snapshots.jackson.serializers.JacksonSnapshotSerializerTest.shouldSerializeDifferentTypes=[
  {
    "anObject": { },
    "aByte": 65,
    "aShort": 32767,
    "anInt": 2147483647,
    "aLong": 9223372036854775807,
    "aBigDecimal": 100,
    "aFloat": 0.1234567,
    "aDouble": 1.123456789123456,
    "aBoolean": true,
    "aChar": "A",
    "string": "Hello World",
    "date": "2020-10-19T22:21:07.103+00:00",
    "localDate": "2020-10-19",
    "localDateTime": "2020-10-19T22:21:07.103",
    "zonedDateTime": "2020-04-19T22:21:07.103+10:00[Australia/Melbourne]",
    "anEnum": "A",
    "presentOptional": "Hello World",
    "emptyOptional": null,
    "stringArray": [
      "f",
      "a",
      "d",
      "e",
      "g",
      "b",
      "c"
    ],
    "anEnumArray": [
      "F",
      "A",
      "D",
      "E",
      "G",
      "B",
      "C"
    ],
    "hashMap": {
      "a": 97,
      "b": 98,
      "c": 99,
      "d": 100,
      "e": 101,
      "f": 102,
      "g": 103
    },
    "treeMap": {
      "a": 97,
      "b": 98,
      "c": 99,
      "d": 100,
      "e": 101,
      "f": 102,
      "g": 103
    },
    "linkedHashMap": {
      "a": 97,
      "b": 98,
      "c": 99,
      "d": 100,
      "e": 101,
      "f": 102,
      "g": 103
    },
    "linkedHashSet": [
      "f",
      "a",
      "d",
      "e",
      "g",
      "b",
      "c"
    ],
    "hashSet": [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
      "g"
    ],
    "treeSet": [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
      "g"
    ],
    "arrayList": [
      "f",
      "a",
      "d",
      "e",
      "g",
      "b",
      "c"
    ],
    "linkedList": [
      "f",
      "a",
      "d",
      "e",
      "g",
      "b",
      "c"
    ],
    "listOfCollections": [
      {
        "a": 97,
        "b": 98,
        "c": 99,
        "d": 100,
        "e": 101,
        "f": 102,
        "g": 103
      },
      [
        "f",
        "a",
        "d",
        "e",
        "g",
        "b",
        "c"
      ],
      [
        "f",
        "a",
        "d",
        "e",
        "g",
        "b",
        "c"
      ]
    ]
  }
]